// dev-portal-common/migrate-group-membership.js
'use strict'

// The following environment variables must be set appropriately:
// - UserPoolId
// - AdminGroupName
// - RegisteredGroupName

const AWS = require('aws-sdk')
const cognitoIdentityServiceProvider = new AWS.CognitoIdentityServiceProvider({
  apiVersion: '2016-04-18'
})

exports.cognitoIdentityServiceProvider = cognitoIdentityServiceProvider

exports.performMigration = async (username, userGroups) => {
  console.log('Migrating user groups')

  const promises = []
  const groups = userGroups.split(',')

  if (groups.includes('admin')) {
    console.log('Adding user to admin group')
    promises.push(cognitoIdentityServiceProvider.adminAddUserToGroup({
      UserPoolId: process.env.UserPoolId,
      Username: username,
      GroupName: process.env.AdminGroupName
    }).promise())
  }

  if (groups.includes('registered')) {
    console.log('Adding user to registered group')
    promises.push(cognitoIdentityServiceProvider.adminAddUserToGroup({
      UserPoolId: process.env.UserPoolId,
      Username: username,
      GroupName: process.env.RegisteredGroupName
    }).promise())
  }

  console.log('Deleting custom tracking attribute')
  promises.push(cognitoIdentityServiceProvider.adminDeleteUserAttributes({
    UserPoolId: process.env.UserPoolId,
    Username: username,
    UserAttributeNames: ['custom:groups']
  }).promise())

  await Promise.all(promises)
}
