// dev-portal-common/pager.js
/**
 * Returns an array of all items across pages.
 *
 *  - The `fetchPage` function may accept a "fetch parameters" object and must
 *    return a Promise resolving to "page data".
 *  - The `commonParams` object represents "fetch parameters" which are passed
 *    to every `fetchPage` call.
 *  - The `selectItems` function must accept "page data" and must return an
 *    iterable of the page's items.
 *  - The `getNextPageParams` function may accept "page data" and must return
 *    either:
 *    - a "fetch parameters" object which will be merged with `commonParams` on
 *      the next call to `fetchPage`, indicating that there is a next page to
 *      fetch; or
 *    - a falsy value, to indicate that no next page should be fetched.
 */
const fetchAllItems = async ({ fetchPage, commonParams, selectItems, getNextPageParams }) => {
  const firstPage = await fetchPage(commonParams).promise()
  const items = [...selectItems(firstPage)]
  let nextPageParams = getNextPageParams(firstPage)
  while (nextPageParams) {
    const page = await fetchPage({
      ...commonParams,
      ...nextPageParams
    }).promise()
    items.push(...selectItems(page))
    nextPageParams = getNextPageParams(page)
  }
  return items
}

const fetchUsersInCognitoUserPoolGroup = ({ cognitoClient, userPoolId, groupName }) => fetchAllItems({
  fetchPage: params => cognitoClient.listUsersInGroup(params),
  commonParams: { UserPoolId: userPoolId, GroupName: groupName },
  selectItems: page => page.Users,
  getNextPageParams: page => page.NextToken && { NextToken: page.NextToken }
})

const fetchUsersInCognitoUserPool = ({ cognitoClient, userPoolId, extraParams = {} }) => fetchAllItems({
  fetchPage: params => cognitoClient.listUsers(params),
  commonParams: { ...extraParams, UserPoolId: userPoolId },
  selectItems: page => page.Users,
  getNextPageParams: ({ PaginationToken }) => PaginationToken && { PaginationToken }
})

const fetchGroupsForCognitoUser = ({ cognitoClient, userPoolId, username }) => fetchAllItems({
  fetchPage: params => cognitoClient.adminListGroupsForUser(params),
  commonParams: { UserPoolId: userPoolId, Username: username },
  selectItems: page => page.Groups,
  getNextPageParams: ({ NextToken }) => NextToken && { NextToken }
})

const getDynamoDbNextPageParams = page =>
  page.LastEvaluatedKey && { ExclusiveStartKey: page.LastEvaluatedKey }

const fetchItemsInDynamoDbTable = ({ dynamoDbClient, tableName, extraParams = {} }) => fetchAllItems({
  fetchPage: params => dynamoDbClient.scan(params),
  commonParams: { ...extraParams, TableName: tableName },
  selectItems: page => page.Items,
  getNextPageParams: getDynamoDbNextPageParams
})

const fetchApiGatewayApiKeys = ({ apiGatewayClient, identityId, extraParams = {} }) => fetchAllItems({
  fetchPage: params => apiGatewayClient.getApiKeys(params),
  commonParams: { ...extraParams, nameQuery: identityId },
  selectItems: ({ items }) => items,
  getNextPageParams: ({ position }) => position && { position }
})

module.exports = {
  fetchAllItems,
  fetchUsersInCognitoUserPoolGroup,
  fetchUsersInCognitoUserPool,
  fetchItemsInDynamoDbTable,
  fetchGroupsForCognitoUser,
  fetchApiGatewayApiKeys
}
